### PR TITLE
feat: Add Lichess Provider

### LINKED ISSUE #3742: Lichess.org Provider Support
There is currently no provider to support the authentication of Lichess.org users. 

I created a provider that I'll link in a pull request to add a Lichess provider

### PR DESCRIPTION
**Pull request by brandon-kong**
# Submitting Pull Requests
Closes #3742 

## General

 - [x] Make sure you use [semantic commit messages](https://seesparkbox.com/foundry/semantic_commit_messages).
       Examples: `"fix(google): Fixed foobar bug"`, `"feat(accounts): Added foobar feature"`.
 - [x] All Python code must formatted using Black, and clean from pep8 and isort issues.
 - [x] JavaScript code should adhere to [StandardJS](https://standardjs.com).
 - [x] If your changes are significant, please update `ChangeLog.rst`.
 - [x] If your change is substantial, feel free to add yourself to `AUTHORS`.

 ## Provider Specifics

 In case you add a new provider:

- [x] Make sure unit tests are available.
- [x] Add an entry of your provider in `test_settings.py::INSTALLED_APPS` and `docs/installation.rst::INSTALLED_APPS`.
- [x] Add documentation to `docs/providers/<provider name>.rst` and `docs/providers/index.rst` Provider Specifics toctree.
- [ ] Add an entry to the list of supported providers over at `docs/overview.rst`.

### CODE REVIEW

File: allauth/socialaccount/providers/lichess/views.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+                self.profile_url,
+                params={"access_token": token.token},
+                headers={
+                    "Authorization": f"Bearer {token.token}",
+                },
+            )
+        )
+
+        # retrieve email address
+        email_data = (
```
Comment (pennersr) [inline allauth/socialaccount/providers/lichess/views.py:36]: This is an additional call that is only needed if this is true:

```
from allauth.socialaccount.app_settings import QUERY_EMAIL
```

Can you make that optional?

File: allauth/socialaccount/providers/lichess/views.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+            .get_requests_session()
+            .get(
+                self.profile_url,
+                params={"access_token": token.token},
+                headers={
+                    "Authorization": f"Bearer {token.token}",
+                },
+            )
+        )
+
```
Comment (pennersr) [inline allauth/socialaccount/providers/lichess/views.py:34]: Can you please add `raise_for_status()` here, so something like:

```
resp = ...get(...)
resp.raise_for_status()
extra_data = resp.json()
```

File: allauth/socialaccount/providers/lichess/views.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+        # retrieve email address
+        email_data = (
+            get_adapter()
+            .get_requests_session()
+            .get(
```
Comment (pennersr) [inline allauth/socialaccount/providers/lichess/views.py:39]: Can you add `raise_for_status()` here as well.

File: allauth/socialaccount/providers/lichess/provider.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+            username=data.get("username"),
+            email=data.get("email"),
+            first_name=first_name,
+            last_name=last_name,
+        )
+
+    def extract_email_addresses(self, data):
+        ret = []
+        email = data.get("email")
+        verified = data.get("verified") or False
```
Comment (pennersr) [inline allauth/socialaccount/providers/lichess/provider.py:42]: While there seems to be a `verified` flag over at https://lichess.org/api#tag/Account/operation/accountMe it is not documented what that means. It is also not listed at the email endpoint, so we really cannot tell with confidence that this really implies that the email address is verified. I don't think we can rely on this.

**Review by pennersr — CHANGES_REQUESTED** [review]

**Comment by brandon-kong** [comment]
@pennersr Thank you for your feedback. I went ahead and applied the changes you requested.

File: ChangeLog.rst [hunk 5]
```diff
@@ -6,7 +6,7 @@ Note worthy changes
 
 - Added a dummy provider, useful for testing purposes: ``allauth.socialaccount.providers.dummy``.
 
-- Added a new provider, Atlassian
+- Added a new provider, Atlassian, Lichess
```
Comment (pennersr) [inline ChangeLog.rst:9]: Can you put this below the 0.63.0 heading?

File: allauth/socialaccount/providers/lichess/views.py [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+)
+
+
+class LichessOAuth2Adapter(OAuth2Adapter):
+    provider_id = "lichess"
+
+    settings = app_settings.PROVIDERS.get(provider_id, {})
+    provider_base_url = settings.get("API_URL", "https://lichess.org")
+
+    access_token_url = "{0}/api/token".format(provider_base_url)
```
Comment (pennersr) [inline allauth/socialaccount/providers/lichess/views.py:17]:
```suggestion
    access_token_url = f"{provider_base_url}/api/token"
```

File: allauth/socialaccount/providers/lichess/views.py [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+    authorize_url = "{0}/oauth".format(provider_base_url)
```
Comment (pennersr) [inline allauth/socialaccount/providers/lichess/views.py:18]:
```suggestion
    authorize_url = f"{provider_base_url}/oauth"
```

File: allauth/socialaccount/providers/lichess/views.py [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+
+    profile_url = "{0}/api/account".format(provider_base_url)
+    email_address_url = "{0}/api/account/email".format(provider_base_url)
```
Comment (pennersr) [inline allauth/socialaccount/providers/lichess/views.py:21]:
```suggestion
    email_address_url = f"{provider_base_url}/api/account/email"
```

File: allauth/socialaccount/providers/lichess/views.py [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+class LichessOAuth2Adapter(OAuth2Adapter):
+    provider_id = "lichess"
+
+    settings = app_settings.PROVIDERS.get(provider_id, {})
+    provider_base_url = settings.get("API_URL", "https://lichess.org")
+
+    access_token_url = "{0}/api/token".format(provider_base_url)
+    authorize_url = "{0}/oauth".format(provider_base_url)
+
+    profile_url = "{0}/api/account".format(provider_base_url)
```
Comment (pennersr) [inline allauth/socialaccount/providers/lichess/views.py:20]:
```suggestion
    profile_url = f"{provider_base_url}/api/account"
```

**Review by pennersr — CHANGES_REQUESTED** [review]

**Comment by pennersr** [comment]
Merged via 28e733bc

**Pull request closed without merging**